src/java/org/apache/cassandra/service/ReplicaFilteringProtection.java

```diff
@@ -132,34 +147,33 @@
      */
     UnfilteredPartitionIterator queryProtectedPartitions(int source)
     {
-        UnfilteredPartitionIterator original = makeIterator(originalPartitions.get(source));
-        SortedMap<DecoratedKey, BTreeSet.Builder<Clustering>> toFetch = rowsToFetch.get(source);
+        UnfilteredPartitionIterator original = makeIterator(originalPartitions.set(source, null));
+        SortedMap<DecoratedKey, BTreeSet.Builder<Clustering>> toFetch = rowsToFetch.set(source, null);
 
         if (toFetch.isEmpty())
             return original;
 
-        // TODO: this would be more efficient if we had multi-key queries internally
+        // TODO: This would be more efficient if we had multi-key queries internally (see CASSANDRA-15910)
         List<UnfilteredPartitionIterator> fetched = toFetch.keySet()
                                                            .stream()
-                                                           .map(k -> querySourceOnKey(source, k))
+                                                           .map(k -> querySourceOnKey(source, k, toFetch.get(k)))
                                                            .collect(Collectors.toList());
```

Comment:
@maedhroz Given that we don't have multi-key queries, we can easily reduce the number RFP queries in some cases if we use an iterator instead of building a list:
```java
Iterator<UnfilteredPartitionIterator> fetched = toFetch.keySet()
                                                       .stream()
                                                       .map(k -> querySourceOnKey(source, k, toFetch.get(k)))
                                                       .iterator();
```
For example, the following test would go down from four RFP queries to just one, although still needing four SRP requests:
```python
self._prepare_cluster(
    create_table="CREATE TABLE t (k int PRIMARY KEY, v text)",
    create_index="CREATE INDEX ON t(v)",
    both_nodes=["INSERT INTO t (k, v) VALUES (1, 'old')"],
    only_node1=["UPDATE t SET v = 'new' WHERE k = 1",
                "UPDATE t SET v = 'new' WHERE k = 2",
                "UPDATE t SET v = 'new' WHERE k = 3",
                "UPDATE t SET v = 'new' WHERE k = 4"])
self._assert_one("SELECT * FROM t WHERE v = 'new' LIMIT 1", row=[1, 'new'])
```
Changing to an iterator would also require modifying the method `UnfilteredPartitionIterator.concat` to accept an iterator instead of a list:
```java
public static UnfilteredPartitionIterator concat(final Iterator<UnfilteredPartitionIterator> iterators)
{
    assert iterators.hasNext();
    UnfilteredPartitionIterator first = iterators.next();
    if (!first.hasNext())
        return first;
    class Extend implements MorePartitions<UnfilteredPartitionIterator>
    {
        public UnfilteredPartitionIterator moreContents()
        {
            return iterators.hasNext() ? iterators.next() : null;
        }
    }
    return MorePartitions.extend(first, new Extend());
}
```

Reply:
@adelapena Indeed. Let's be lazy.